plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -103,58 +54,47 @@ not have side effects; (2) all arguments be pure, since otherwise it is unsafe t
 perform beta reduction.
 -}
 
-{- | Computes the 'Utils.Arity' of a term. Also returns the function body, for checking whether
-it's `Utils.acceptable`.
--}
-splitParams ::
-  Term tyname name uni fun ann ->
-  (Arity tyname name, Term tyname name uni fun ann)
-splitParams = \case
-  LamAbs _ n _ body ->
-    let (nextArgs, nextBody) = splitParams body in (TermParam n : nextArgs, nextBody)
-  TyAbs _ n _ body ->
-    let (nextArgs, nextBody) = splitParams body in (TypeParam n : nextArgs, nextBody)
-  -- Whenever we encounter a body that is not a lambda or type abstraction, we are done counting
-  tm -> ([], tm)
-
-{- | Fully apply the RHS of the given variable to the given arguments, and beta-reduce
+{- | Apply the RHS of the given variable to the given arguments, and beta-reduce
 the application, if possible.
 -}
-fullyApplyAndBetaReduce ::
+applyAndBetaReduce ::
   forall tyname name uni fun ann.
   (InliningConstraints tyname name uni fun) =>
-  -- | The variable
-  VarInfo tyname name uni fun ann ->
+  -- | The rhs of the variable, should have been renamed already
+  Term tyname name uni fun ann ->
   -- | The arguments
   AppContext tyname name uni fun ann ->
   InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
-fullyApplyAndBetaReduce info args0 = do
-  rhsBody <- liftDupable (let Done rhsBody = varRhsBody info in rhsBody)
+applyAndBetaReduce rhs args0 = do
   let go ::
+        -- | The rhs of the variable, should have been renamed already
         Term tyname name uni fun ann ->
-        Arity tyname name ->
         AppContext tyname name uni fun ann ->
         InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
-      go acc arity args = case (arity, args) of
-        -- success
-        ([], _) -> pure . Just $ fillAppContext acc args
-        (TermParam param : arity', TermAppContext arg _ args') -> do
-          safe <- safeToBetaReduce param arg
-          if safe
+      go acc args = case (acc, args) of
+        (LamAbs _ann n _ty tm, appCtx@(TermAppContext arg _ args')) -> do
+          safe <- safeToBetaReduce n arg acc
+          if safe -- we only do substitution if it is safe to beta reduce
             then do
-              acc' <-
-                termSubstNamesM
-                  (\n -> if n == param then Just <$> PLC.rename arg else pure Nothing)
-                  acc
-              go acc' arity' args'
-            else pure Nothing
-        (TypeParam param : arity', TypeAppContext arg _ args') -> do
+              acc' <- do
+                termSubstNamesM -- substitute the term param with the arg in the function body
+                  -- rename before substitution to ensure global uniqueness
+                  (\tmName -> if tmName == n then Just <$> PLC.rename arg else pure Nothing)
```

Comment:
This is actually another example: do we need to rename here? If the input is globally unique, then I think this will just give us something with local uniqueness... which is fine so long as we then rename it before we substitute it into the original program.

OTOH this reasoning is a bit tricky and maybe it's better to just stick on the safe side and rename too much.